### DISCUSSION TITLE
Bionic Enterprise + Ollama with ROCm?

### DISCUSSION THREAD

**Discussion by isensez** [bionic-gpt/bionic-gpt#369] [Q&A] [answered]
Hello Bionic community,

Is it possible or has anyone been able to get the enterprise version of Bionic in K8s connect to a local Ollama? Anyone tried this with Ollama ROCm on AMD GPUs? 
I have my own Ollama version compiled for ROCm supoort and its working as intended on my setup. Bionic Enterprise is working perfectly out of the box, but i have yet to be successful getting Bionic to interact with the local Ollama instance (running as systemd process, not docker)
Any tips or ideas?

**Comment by ianpurton** [comment]
So the Ollama is running outside the K8's cluster?

So you want to tunnel out of K8's (kind) and access this model?

**Reply by isensez** [reply]
Correct, I have a enterprise Bionic deployment an an external Ollama instance (in the same server and another in a separate server), i want Bionic to be able to reach Ollama locally/externally to access the models i have there.
Is this achievable?

**Reply by ianpurton** [reply, marked as answer]
You need to look here https://bionic-gpt.com/docs/administration/add-a-model/

Set the model name to the same as ollama, set the remote address i.e. http://123.123.123.123/v1

And let me know how that goes.

**Reply by isensez** [reply]
Thank you for the suggestion, i will give it a try and provide feedaback as soon as possible.
Why do we set /v1 as the endpoint? Dont we need to set the api endpoint and port of ollama as specified by their docs? (/chat etc) 
Or /v1 uses the llm-api for conversion?

**Reply by ianpurton** [reply]
We always use the Open AI endpoints https://platform.openai.com/docs/api-reference

Just recently ollama starting supporting them. https://registry.ollama.ai/blog/openai-compatibility

The v1 prefix lines up with Open AI API. But, to be honest causes more confusion than anything else.

Let me know how it goes.

**Reply by isensez** [reply]
After some testing i have good and bad news.
The good:
Successfully works connected to Ollama locally in the same machine and can use the different models in Ollama.

The bad:
Seems that Ollama doesnt reset when being prompted for a different model after using one, the response always fails with "
Error occurred while generating." when switching prompts that use a different model.
If i restart the Ollama systemd process i am able to load any model and it instantly works for prompting, so restarting again and loading a different model works every time. 
But loading a model, prompting, then changing the prompt to a different model from the first prompt always errors out, and even switching to the original model that was initially loaded also fails.

I specifically tested both models with this setup and they work if they are being individually prompted from the getgo. But switching prompts that use a different model is not working as expected and makes it so any model (except the default llama2) always fail after switching.

This would be great to look into as Ollama facilitates so many aspects of using AMD GPUs, pulling and deleting models etc.

Heres what i did:
Cloned ollama repo
Recompiled binary with ROCm tags for ROCm support (im using a AMD 7900 XTX, passed through to a VM in Proxmox)
Started Ollama and downloaded mistral and mixtral models.

In the Bionic Web UI i setup the model like this:

mistral:latest	http://192.168.120.3:11434/v1	Large Language Model	7 Billion	2048
mixtral:latest	http://192.168.120.3:11434/v1	Large Language Model	7 Billion	2048

**Reply by isensez** [reply]
Further developments:
I managed to get models to switch mid chat, but i had to wait a while before switching models.

Using Mistral in the previous chat, started a new chat and prompted a different model (Mixtral 8x7b), initial response failed, i prompted it again, and it worked.

Switching models seems hit or miss at the moment.

**Reply by ianpurton** [reply]
You could run an instance of ollama for each model. That might fix it.

**Reply by ianpurton** [reply]
So let's get this straight. You're running 8x7b on AMD GPU's? Cool.

Can you tell me the hardware you're using?

**Reply by isensez** [reply]
That will be a bit problematic as i have a dozen models, but would work yes,
Ideally the unloading of the previous model, and loading of the new should be seamless but atleast this is a nice start.
Im not entirely sure what is happening in the background for sometimes working and others failing.

**Reply by isensez** [reply]
Correct, i am using 8x7b Q4, almost entirely fits in memory, can get 11-14 T/s on a VM with the 7900 XTX
HW:
Asrock EPYCD8-2T
AMD EPYC 7452 32-Core CPU (Only assigned and pined 10 cores - 20 threads to the Bionic/Ollama VM)
128GB 3200 MTs (Only assigned 32GB to the VM)
AMD 7900 XTX (passed through to the Bionic/Ollama VM
The storage itself is not local, its in a SSD SANs that plugs via 10G networking to Proxmox

Proxmox as hypervisor.

**Reply by ianpurton** [reply]
Thanks. One thing you should be aware of is ollama doesn't support continuous batching.

So if you have multiple users they will have to wait in a queue to get results. They won't get results in parallel.

For continuous batching look at tgi or vllm.

**Reply by isensez** [reply]
Very important detail and really appreciate the heads up, thank you!
